bootloader/gfx/gfx.c

```diff
@@ -187,7 +187,11 @@ void gfx_putc(char c)
 
 			for (u32 i = 0; i < 16; i += 2)
 			{
-				u8 v = *cbuf++;
+				u8 v;
+				//every first(or last depending on your view) line of bitmap MUST be 0 to prevent artefacts from preceeding char
+				//without this, last (or first) u8 of previous char gets pulled and rendered once
+				if (i != 0) v = *cbuf++;
+				else v = 0x00;
```

Comment:
Yes. Thats precisely what I did.

Try removing it and you will see it is a line "off" if using any other font.

The previous char "last lines" are rendered as part of the new desired character.

This just zeroes the last / first row, as it was getting rendered twice on different characters. Not noticeable on default Hekate font.

Reply:
You tried the code above and it still does it?
I'll need to check it again if yes.

Reply:
Tried code above. Completely screws the rendering up.
Now, with the above code, in a 8x8 char(well, 16 x 16 as we are doubling)

(1-8)x16 is desired char,
(9-16) is prev char.

ie:

hekate becomes:

hekate
iflbuf

Reply:
Hmm ok, that's strange. Basically the j loop does 8 + 8 px in line and the k loop paints the second line. So it should not break like you describe. Because with the 2 line change I posted it should paint the 2 16px lines and then change cbuf. The 1-8 you talked about, is 1st pixel to 8th pixel in the line?

Example:
correct char: -
incorrect char: =

One `i` loop with correct output:
```
-- // `j` loop
-- // `j` loop from 2nd iteration of `k` loop
```

So it's like
```
--
==
```
or
```
-=
-=
```

I suppose your branch exhibits both issues (master vs code I linked) with the font you use.
Then I can haunt it down.

Reply:
I think what is happening, is for some reason, the alignment is "off by one" line.

When I try the above code you posted, it renders the u8`s from the previous character array.

To minimise the effect, I first tried simply subtracting...
`u8 *cbuf = (u8 *)&_gfx_font[8 * (c - 32) - 1];`

This helped negate it somewhat, but on characters in any font that cover the very last row (for example "g"), the last row of g would be rendered.

It is as though there is one single line of "space" missing.

The only workaround I could make work correctly and maintain spacing for any font, was to add the blank u8 (0x00) everytime i=0.

You would think this would chop off the first line of the character in both iterations, but it doesn`t...

Reply:
The font works like that:
Every byte is a line. Every 8 bytes it's a char.

And because it prints the prev char for you, it means that it should be this:
`u8 *cbuf = (u8 *)&_gfx_font[8 * (c - 32) + 8];` and not `-8`.

If your font proper? ascii no == char?

print all chars

```c
gfx_printf(" !"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHI\nJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqr\nstuvwxyz{|}~\n");
```

Reply:
Yep its proper. Tried with at least a dozen. Just exported as a C include.

On closer examination, it is only happening on one iteration.

Reply:
What do you mean `on one iteration`?

Reply:
The j loop is repeated twice. On the second iteration (controlled by k, that renders the bottom half), the very last rendered row of the next char in the array.

To see the bug, try this...

We will use the char "h" in this example... Char 104.
In the following char, which is "i"(char 105), substitute the first empty line (0x00) with a solid black line (0xFF). So the first row of "i" becomes a solid black line.

Compile, run.

You will see your "h" have a solid black line at the bottom, on the second iteration.

Hope this helps!

Reply:
https://github.com/CTCaer/hekate/commit/1ccce5f1a269635dcb76f8620d10b6ad65cdea90

This fixes the off by one in the horizontal line doubling.
Programmatically the code is now correct and by using a full sized 8x8 px font, the results are also correct.

Anything else is probably wrong porting of the font. Even before, the only way to render a part of **next** char was the last vertical line of the 2nd horizontal pass.